modules/optimization/src/executioners/AdjointSolve.C

```diff
@@ -0,0 +1,163 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdjointSolve.h"
+#include "OptimizationAppTypes.h"
+
+#include "FEProblem.h"
+#include "NonlinearSystemBase.h"
+#include "NonlinearSystem.h"
+#include "NodalBCBase.h"
+
+#include "libmesh/petsc_matrix.h"
+#include "libmesh/petsc_vector.h"
+#include "petscmat.h"
+
+InputParameters
+AdjointSolve::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  params.addRequiredParam<NonlinearSystemName>(
+      "forward_system", "Name of the system representing the forward problem.");
+  params.addRequiredParam<NonlinearSystemName>(
+      "adjoint_system", "Name of the system representing the adjoint problem.");
+  return params;
+}
+
+AdjointSolve::AdjointSolve(Executioner & ex)
+  : SolveObject(ex),
+    _forward_sys_num(_problem.nlSysNum(getParam<NonlinearSystemName>("forward_system"))),
+    _adjoint_sys_num(_problem.nlSysNum(getParam<NonlinearSystemName>("adjoint_system"))),
+    _nl_forward(_problem.getNonlinearSystemBase(_forward_sys_num)),
+    _nl_adjoint(_problem.getNonlinearSystemBase(_adjoint_sys_num))
+{
+  // These should never be hit, but just in case
+  if (!dynamic_cast<NonlinearSystem *>(&_nl_forward))
+    paramError("forward_system", "Forward system does not appear to be a 'NonlinearSystem'.");
+  if (!dynamic_cast<NonlinearSystem *>(&_nl_adjoint))
+    paramError("adjoint_system", "Adjoint system does not appear to be a 'NonlinearSystem'.");
+}
+
+bool
+AdjointSolve::solve()
+{
+  TIME_SECTION("execute", 1, "Executing adjoint problem", false);
+
+  if (_inner_solve && !_inner_solve->solve())
+    return false;
+
+  _problem.execute(OptimizationAppTypes::EXEC_ADJOINT_TIMESTEP_BEGIN);
+  if (!_problem.execMultiApps(OptimizationAppTypes::EXEC_ADJOINT_TIMESTEP_BEGIN))
+  {
+    _console << "MultiApps failed to converge on ADJOINT_TIMESTEP_BEGIN!" << std::endl;
+    return false;
+  }
+
+  checkIntegrity();
+
+  // Convenient references
+  // Adjoint matrix, solution, and right-hand-side
+  auto & matrix = dynamic_cast<ImplicitSystem *>(&_nl_forward.system())->get_system_matrix();
```

Comment:
If you know this is an implicit system, then you can use `static_cast` here (and in following lines). Otherwise we should do a check to make sure this cast is non-null

Reply:
I have error checks in the constructor. So I'll switch to `static_cast`